lib/handleRequest.js

```diff
@@ -89,14 +89,45 @@ function preValidationCallback (err, request, reply) {
     return
   }
 
-  const result = validateSchema(reply[kRouteContext], request)
-  if (result) {
+  const validationErr = validateSchema(reply[kRouteContext], request)
+  const isAsync = validationErr && typeof validationErr.then === 'function'
+
+  if (isAsync) {
+    const cb = asyncValidationCompleted.bind(null, request, reply)
+    validationErr.then(cb, cb)
+  } else {
+    if (validationErr) {
+      if (reply[kRouteContext].attachValidation === false) {
+        reply.send(validationErr)
+        return
+      }
+
+      reply.request.validationError = validationErr
+    }
+
+    // preHandler hook
+    if (request[kRouteContext].preHandler !== null) {
+      hookRunner(
+        request[kRouteContext].preHandler,
+        hookIterator,
+        request,
+        reply,
+        preHandlerCallback
+      )
+    } else {
+      preHandlerCallback(null, request, reply)
+    }
```

Comment:
Looking at the code, why can't we just call `asyncValidationCompleted` here?

Reply:
Done - I was focusing on avoiding the callback 😂